parsec/backend/memory/pki.py

```diff
@@ -0,0 +1,153 @@
+                        f"Certificate {str(certificate_id)} already attributed",
+                    )
+
+                if not force_flag:
+                    raise PkiCertificateAlreadyRequestedError(
+                        existing_certificate.reply_timestamp,
+                        f"Certificate {str(certificate_id)} already used in request {request_id}",
+                    )
+
+        new_pki_certificate = PkiCertificate(
+            certificate_id=certificate_id,
+            request_id=request_id,
+            request_timestamp=pendulum.now(),
+            request_object=request_object,
+            reply_user_id=None,
+            reply_timestamp=None,
+            reply_object=None,
+        )
+        self._pki_certificates[certificate_id] = new_pki_certificate
+        return new_pki_certificate.request_timestamp
+
+    async def pki_enrollment_get_requests(self) -> List[Tuple[bytes, UUID, PkiEnrollmentRequest]]:
+        return [
+            (
+                pki_certificate.certificate_id,
+                pki_certificate.request_id,
+                pki_certificate.request_object,
+            )
+            for pki_certificate in self._pki_certificates.values()
+            if pki_certificate.reply_admin_user is None
+        ]
+
+    async def pki_enrollrment_reply_reject_user(
+        self, pki_reply_bundle: PkiEnrollementReplyBundle
+    ) -> DateTime:
+        return await self._pki_enrollment_reply(pki_reply_bundle)
+
+    async def pki_enrollment_reply_approve_user(
+        self, pki_reply_bundle: PkiEnrollementReplyBundle, client_ctx, msg
+    ) -> DateTime:
+        await self._user_component._api_user_create(client_ctx, msg, pki_reply_bundle)
+        timestamp = self._pki_certificates[pki_reply_bundle.certificate_id].reply_timestamp
+        assert timestamp
+        return timestamp
+
+    async def _pki_enrollment_reply(self, pki_reply_bundle: PkiEnrollementReplyBundle) -> DateTime:
+        try:
+            pki_certificate = self._pki_certificates[pki_reply_bundle.certificate_id]
+        except KeyError:
+            raise PkiCertificateNotFoundError(
+                f"Certificate {str(pki_reply_bundle.certificate_id)} not found"
+            )
+        if pki_certificate.request_id != pki_reply_bundle.request_id:
+            raise PkiCertificateRequestNotFoundError(
+                f"Request {pki_reply_bundle.request_id} not found for certificate {str(pki_reply_bundle.certificate_id)}"
+            )
+
+        pki_certificate.reply_object = pki_reply_bundle.reply_object
+        pki_certificate.reply_user_id = pki_reply_bundle.reply_user_id
+        pki_certificate.reply_timestamp = pendulum.now()
+        pki_certificate.reply_admin_user = pki_reply_bundle.reply_admin_user
+        return pki_certificate.reply_timestamp
+
+    async def pki_enrollment_get_reply(
+        self, certificate_id, request_id
+    ) -> Tuple[
+        Optional[PkiEnrollmentReply],
```

Comment:
Too much optional here, the combinatorics is very high then !

I have the feeling this is in fact much simpler: either a reply is present and all field are here, either not yet and we only have the request timestamp !

but it is weird to return the request timestamp given the name of the method :(

So I would rename it `get_enrollment_state()` and it would return a `PkiEnrollementStateRequest` or `PkiEnrollementStateReply` (with both being a simple dataclass)